src/Compilers/VisualBasic/Portable/Symbols/AssemblySymbol.vb

```diff
@@ -534,45 +534,73 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
             result = Nothing
 
-            If includeReferences Then
-                ' Lookup in references
-                Dim references As ImmutableArray(Of AssemblySymbol) = Me.Modules(0).GetReferencedAssemblySymbols()
+            If Not includeReferences Then
+                Return result
+            End If
 
-                For i As Integer = 0 To references.Length - 1 Step 1
-                    Debug.Assert(Not (TypeOf Me Is SourceAssemblySymbol AndAlso references(i).IsMissing)) ' Non-source assemblies can have missing references
-                    Dim candidate As NamedTypeSymbol = references(i).LookupTopLevelMetadataType(metadataName, digThroughForwardedTypes:=False)
+            ' Then try corlib, when finding a result there means we've found the final result
+            Dim skipCorLibrary = False
 
-                    If isWellKnownType AndAlso Not IsValidWellKnownType(candidate) Then
-                        candidate = Nothing
-                    End If
+            If CorLibrary IsNot Me AndAlso
+                Not CorLibrary.IsMissing AndAlso
+                Not ignoreCorLibraryDuplicatedTypes Then
 
-                    If IsAcceptableMatchForGetTypeByNameAndArity(candidate) AndAlso
+                Dim corLibCandidate As NamedTypeSymbol = CorLibrary.LookupTopLevelMetadataType(metadataName, digThroughForwardedTypes:=False)
+                skipCorLibrary = True
+
+                If corLibCandidate IsNot Nothing AndAlso
+                    (Not isWellKnownType OrElse IsValidWellKnownType(corLibCandidate)) AndAlso
+                    IsAcceptableMatchForGetTypeByNameAndArity(corLibCandidate) AndAlso
+                    Not corLibCandidate.IsHiddenByVisualBasicEmbeddedAttribute() AndAlso
+                    Not corLibCandidate.IsHiddenByCodeAnalysisEmbeddedAttribute() Then
```

Comment:
Consider extracting these checks and reusing in loop below.

Reply:
It is a bit of repetition. I think I still prefer this way though, especially in absence of local functions for VB...

---
In reply to: [586787930](https://github.com/dotnet/roslyn/pull/51571#discussion_r586787930) [](ancestors = 586787930)